### PR TITLE
feat!: Require Node.js `^18.18.0 || ^20.9.0 || >=21.1.0`

### LINKED ISSUE #17595: Change Request: Drop Node v12/v14/v16/v17/v19 support
### ESLint version

8.49.0

### What problem do you want to solve?

ESLint v8 supports `"node": "^12.22.0 || ^14.17.0 || >=16.0.0"`.

Node 12, 14, 16, 17, 19 are EOL.

* Node 12 EOL: 2022-04-30
* Node 14 EOL: 2023-04-30
* Node 16 EOL: 2023-09-11 (https://nodejs.org/en/blog/announcements/nodejs16-eol)
* Node 17 EOL: 2022-06-01
* Node 19 EOL: 2023-06-01

### What do you think is the correct solution?

In v9, we'd drop support for EOL versions and keep only Node 18, Node 20, and above.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

At the moment, I don't have a concrete proposal for minor versions of Node 18 and Node 20 we'd like to support. Any suggestions?

### PR DESCRIPTION
**Pull request by mdjermanovic**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

Fixes #17595

#### What changes did you make? (Give an overview)

Updated `package.json` to require:

```
"node": "^18.18.0 || ^20.9.0 || >=21.1.0"
```

This drops support for Node.js 12/14/16/17/19.

#### Is there anything you'd like reviewers to focus on?

Did I miss any places that should be updated?

### CODE REVIEW

**Comment by mdjermanovic** [comment]
This is ready for review.

**Comment by ota-meshi** [comment]
What do you think about using `Intl.Segmenter` instead of `graphemer` package?

https://github.com/eslint/eslint/issues/17595#issuecomment-1780185430

**Comment by mdjermanovic** [comment]
> What do you think about using `Intl.Segmenter` instead of `graphemer` package?
> 
> [#17595 (comment)](https://github.com/eslint/eslint/issues/17595#issuecomment-1780185430)

Makes sense to me, though I'm not sure if we should make changes to the `key-spacing` rule since it's now deprecated.

Would it be 100% functionally equivalent?

**Comment by ota-meshi** [comment]
I think it is better to reduce the number of dependency packages even if the rules that affect them are deprecated, since it is beneficial for ESLint to reduce the number of dependency packages. What do you think?

> Would it be 100% functionally equivalent?

Hmm. It seems that they are not 100% equivalent.
However, I think the purpose of both is the same. I think it would be accepted as a minor breaking change in a major version update, what do you think?

**Review by nzakas — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -45,7 +45,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        node: [21.x, 20.x, 19.x, 18.x, 17.x, 16.x, 14.x, 12.x, "12.22.0"]
+        node: [21.x, 20.x, 18.x, "18.18.0"]
```
Comment (kibertoad) [inline .github/workflows/ci.yml:48]: why not 18.x?

**Comment by mdjermanovic** [comment]
> Let's spin off the `Intl.Segmenter` discussion into a separate issue so we can continue discussing.

Agreed. @ota-meshi would you like to open an issue to discuss replacing `graphemer` with `Intl.segmenter`?

**Comment by ota-meshi** [comment]
> Agreed. @ota-meshi would you like to open an issue to discuss replacing graphemer with Intl.segmenter?

I'll do it later when I have time.
If possible, I would like to investigate and show what kind of characters there are differences.

**Review by aladdin-add — APPROVED** [review]
I noticed several things can be updated:

* https://github.com/eslint/eslint/blob/0dd9704c4751e1cd02039f7d6485fee09bbccbf6/lib/config/flat-config-schema.js#L12-L16
* https://github.com/eslint/eslint/blob/0dd9704c4751e1cd02039f7d6485fee09bbccbf6/lib/eslint/flat-eslint.js#L12-L13

sure, it's not a blocker of merging this one, we can update it in another PR later.

File: package.json [hunk 2]
```diff
@@ -170,6 +170,6 @@
   ],
   "license": "MIT",
   "engines": {
-    "node": "^12.22.0 || ^14.17.0 || >=16.0.0"
+    "node": "^18.18.0 || ^20.9.0 || >=21.1.0"
```
Comment (koddsson) [inline package.json:173]: Hey sorry for the random comment but why `^18.18.0` and not `^18.0.0`? I've been trying to dig around for this and can't find the reason. I've got the node engine set to `>=18` in my app so I'd need to upgrade which is fine, I'm just curious :)

Comment (kibertoad) [reply package.json:173]: It's the first LTS release for 18

Comment (Tanujkanti4441) [reply package.json:173]: have you went through the discussion in the related [issue](https://github.com/eslint/eslint/issues/17595)?

Comment (koddsson) [reply package.json:173]: > have you gone through the discussion in the related [issue](https://github.com/eslint/eslint/issues/17595)?

Yeah but I'm missing why those versions were chosen. The only thing I see is it being the latest versions at the time.

**Pull request merged**